packages/common/src/location/url_service.ts

```diff
@@ -0,0 +1,172 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Injectable, ɵɵinject} from '@angular/core';
+
+import {LocationStrategy} from './location_strategy';
+import {joinWithSlash, normalizeQueryParams, stripTrailingSlash} from './util';
+
+/**
+ * @description
+ *
+ * A service that applications can use to interact with a browser's URL.
+ *
+ * This service depends on `Location` to normalize and update the URL.
+ * @usageNotes
+ *
+ * It's better to use the `Router.navigate()` service to trigger route changes. Use
+ * `UrlService` only if you need to interact with or create normalized URLs outside of
+ * routing.
+ *
+ * `UrlService` is responsible for normalizing the URL against the application's base href.
+ * A normalized URL is absolute from the URL host, includes the application's base href, and has no
+ * trailing slash:
+ * - `/my/app/user/123` is normalized
+ * - `my/app/user/123` **is not** normalized
+ * - `/my/app/user/123/` **is not** normalized
+ *
+ * ### Example
+ *
+ * <code-example path='common/location/ts/path_location_component.ts'
+ * region='LocationComponent'></code-example>
+ */
+@Injectable({
+  providedIn: 'root',
+  // See #23917
+  useFactory: createUrl,
+})
```

Comment:
```suggestion
@Injectable({providedIn: 'root'})
```
You can remove the factory function as well. I don't know the full history of what was happening at the time these factory functions were created, but I know you don't need them anymore.